Add doc for verify mtls among apps in ambient (#15594)
istio/istio.io: content/en/docs/ambient/usage/verify-mtls-enabled/index.md

---
title: Verify mutual TLS is enabled
description: Understand how to verify mTLS is enabled among workloads in an ambient mesh.
weight: 15
owner: istio/wg-networking-maintainers
test: no
---

> Once you have added applications to an ambient mesh, you can easily validate mTLS is enabled among your workloads in the ambient mesh, using one or more of the methods below:

Reviewer: the suggestion has 2 spaces between workloads and using

{{< text syntax=bash >}}
$ istioctl ztunnel-config workloads
NAMESPACE    POD NAME                                IP         NODE                     WAYPOINT PROTOCOL
default      details-v1-857849f66-ft8wx              10.42.0.5  k3d-k3s-default-agent-0  None     HBONE
default      kubernetes                              172.20.0.3                          None     TCP
default      productpage-v1-c5b7f7dbc-hlhpd          10.42.0.8  k3d-k3s-default-agent-0  None     HBONE
default      ratings-v1-68d5f5486b-b5sbj             10.42.0.6  k3d-k3s-default-agent-0  None     HBONE
default      reviews-v1-7dc5fc4b46-ndrq9             10.42.1.5  k3d-k3s-default-agent-1  None     HBONE
default      reviews-v2-6cf45d556b-4k4md             10.42.0.7  k3d-k3s-default-agent-0  None     HBONE
default      reviews-v3-86cb7d97f8-zxzl4             10.42.1.6  k3d-k3s-default-agent-1  None     HBONE
{{< /text >}}

Having HBONE configured on your workload doesn't mean your workload will accept reject any plaintext traffic. If you want your workload to reject plaintext traffic, setting a `PeerAuthentication` policy with mTLS mode set to `STRICT` for your workload.

### Validate mTLS from metrics

> If you have [installed Prometheus](docs/ops/integrations/prometheus/#installation), you can setup port-forwarding for Prometheus by using the following command:

Reviewer 2: ```suggestion
If you have [installed Prometheus](docs/ops/integrations/prometheus/#installation), you can setup port-forwarding by using the following command:
```

Reviewer: This does more than just setting up the port-forward iirc. In other doc we say something like "opens the Prometheus UI" which it may make sense to stick with here as well.

Author: Updated, thanks!

{{< text syntax=bash >}}
$ istioctl dashboard prometheus
{{< /text >}}

> View the values for the TCP metrics in the Prometheus browser window. Select Graph. Enter the `istio_tcp_connections_opened_total` metric or `istio_tcp_connections_closed_total` or `istio_tcp_received_bytes_total` or `istio_tcp_sent_bytes_total` and select Execute. The table displayed in the Console tab includes entries similar as below:

Reviewer 2: ```suggestion
In Prometheus, you can view the values for the TCP metrics. First, select Graph and enter the a metric such as: `istio_tcp_connections_opened_total`, `istio_tcp_connections_closed_total`, `istio_tcp_received_bytes_total`, or `istio_tcp_sent_bytes_total`. Lastly, click Execute. The data will contain entries such as:
```

{{< text syntax=text >}}
istio_tcp_connections_opened_total{
  app="ztunnel", 
  connection_security_policy="mutual_tls", 
  destination_principal="spiffe://cluster.local/ns/default/sa/bookinfo-details", 
  destination_service="details.default.svc.cluster.local", 
  reporter="source", 
  request_protocol="tcp", 
  response_flags="-",
  source_app="sleep", 
  source_principal="spiffe://cluster.local/ns/default/sa/sleep", source_workload_namespace="default"
  ...}
{{< /text >}}

Validate that the `connection_security_policy` value is set to `mutual_tls` along with the expected source and destination identity information.

### Validate mTLS from logs

> You can view either source or destination's ztunnel log to confirm mTLS is enabled, along with peer identities. Below is an example of the source ztunnel's log for a request from the `sleep` service to the `details` service:

Reviewer 2: ```suggestion
You can also view either the source or destination ztunnel log to confirm mTLS is enabled, along with peer identities. Below is an example of the source ztunnel's log for a request from the `sleep` service to the `details` service:
```

{{< text syntax=plain >}}

> 2024-08-21T15:32:05.754291Z	info access	connection complete	src.addr=10.42.0.9:33772 src.workload="sleep-7656cf8794-6lsm4" src.namespace="default"

Reviewer 2: ```suggestion
2024-08-21T15:32:05.754291Z    info access    connection complete    src.addr=10.42.0.9:33772 src.workload="sleep-7656cf8794-6lsm4" src.namespace="default"
```

This should fix the linter

Author: Thanks Daniel! Ian also pinged me, fingers crossed this is it!!

Author: mark this as resolved as i checked in a version fixed this around same time.

```

> Refer to the [verifying ztunnel traffic through logs section](docs/ambient/usage/troubleshoot-ztunnel/#verifying-ztunnel-traffic-through-logs) for more details.

Reviewer: Neither here nor in the linked page do we specifically mention specifically what we want them to look for to validate mTLS is working. We should probably have them check that src/dst identity is populated/correct since this is the most direct tie in to the certs we used.

Author: updated!

### Validate with Kiali dashboard

If you have Kiali and Prometheus installed, you can visualize your workload communication in the ambient mesh using Kiali's dashboard. You can see if the connection between any workloads has the paddlelock icon to validate that mTLS is enabled, along with the peer identity information:

{{< image link="./kiali-mtls.png" caption="Kiali dashboard" >}}

> Refer to the [Visualize the application and metrics](docs/ambient/getting-started/secure-and-visualize/#visualize-the-application-and-metrics) for more details.

Reviewer 2: ```suggestion
Refer to the [Visualize the application and metrics](docs/ambient/getting-started/secure-and-visualize/#visualize-the-application-and-metrics) document for more details.
```

### Validate with tcpdump

> If you have access to your Kubernetes worker nodes, you can run the `tcpdump` command to capture all traffic on the network interface, with optional focusing the application ports and HBONE port, for example, port `9080` is the `details` service port and `15008` is the HBONE port.

Reviewer 2: ```suggestion
If you have access to your Kubernetes worker nodes, you can run the `tcpdump` command to capture all traffic on the network interface, with optional focusing the application ports and HBONE port. In this example, port `9080` is the `details` service port and `15008` is the HBONE port:
```

{{< text syntax=bash >}}
$ tcpdump -nAi eth0 port 9080 or port 15008
{{< /text >}}

You should see encrypted traffic from the output of the `tcpdump` command.

> If you don't have access to the worker nodes, you may be able to use the [netshoot image](https://hub.docker.com/r/nicolaka/netshoot) to easily run the command:

Reviewer 2: ```suggestion
If you don't have access to the worker nodes, you may be able to use the [netshoot container image](https://hub.docker.com/r/nicolaka/netshoot) to easily run the command:
```
